components/selector/__init__.py

```diff
@@ -0,0 +1,475 @@
+        """
+        Perform the selection of the groups based on:
+        * Resource availability
+        * 80% chance of completion (TBD)
+        for the given site(s) and night index.
+
+        For each program, begin at the root group and iterate down to the leaves.
+        Each leaf contains an Observation. Filter out Observations that cannot be performed
+        at one of the given sites.
+
+        For each Observation group node, calculate:
+        1. The minimum required conditions to perform the Observation.
+        2. For each night index:
+           The time slots for which the observation can be performed (based on resource availability and weather).
+        3. The score of the Observation.
+
+        Bubble this information back up to conglomerate it for the parent groups.
+
+        An AND group must be able to perform all of its children.
+
+        TODO: Currently OR groups cannot be handled. They are not supported by OCS and
+        TODO: attempts to include them will result in a NotImplementedException.
+        TODO: Further design work must be done to determine how to score them and how to
+        TODO: determine the time slots at which they can be scheduled.
+
+        As all groups are in a group tree rooted at a program node, his information is returned on a
+        program-by-program basis.
+        """
+        # If no night indices are specified, assume all night indices.
+        if night_indices is None:
+            night_indices = np.arange(len(self.collector.time_grid))
+
+        # If no manual ranker was specified, create the default.
+        if ranker is None:
+            ranker = Ranker(self.collector, night_indices)
+
+        # The night_indices in the Selector and Ranker must be the same.
+        if not np.array_equal(night_indices, ranker.night_indices):
+            raise ValueError(f'The Ranker must have the same night indices as the Selector select method.')
+
+        # Create the structure to hold the mapping fom program ID to its group info.
+        program_calcs = {}
+        for program_id in Collector.get_program_ids():
+            program = Collector.get_program(program_id)
+            if program is None:
+                logging.error(f'Program {program_id} was not found in the Collector.')
+                continue
+
+            # Calculate the group info and put it in the structure if there is actually group
+            # info data inside of it, i.e. feasible time slots for it in the plan.
+            group_info = self._calculate_group(program.root_group, sites, ranker)
+
+            # TODO: There are many different ways how to structure what we can return.
+            # TODO: Right now we return any program that has a schedulable group.
+            # TODO: Possible changes to this:
+            # TODO: 1. Only return programs where the root group is schedulable.
+            # TODO: 2. Only include schedulable groups in the data for each program.
```

Comment:
I think this way is easier for how greedymax works right now. Optimizer should just work with the options to schedule I think